src/server_manager/electron_app/index.ts

```diff
@@ -237,3 +102,101 @@ function getWebAppUrl() {
   console.log('Launching web app from ' + webAppUrlString);
   return webAppUrlString;
 }
+
+function main() {
+  // prevent window being garbage collected
+  let mainWindow: Electron.BrowserWindow;
+
+  // Mark secure to avoid mixed content warnings when loading DigitalOcean pages via https://.
+  electron.protocol.registerStandardSchemes(['outline'], {secure: true});
+
+  const isSecondInstance = app.makeSingleInstance((argv, workingDirectory) => {
+    // Someone tried to run a second instance, we should focus our window.
+    if (mainWindow) {
+      if (mainWindow.isMinimized()) {
+        mainWindow.restore();
+      }
+      mainWindow.focus();
+    }
+  });
+
+  if (isSecondInstance) {
+    app.quit();
+  }
+
+  app.on('ready', () => {
+    const menuTemplate = menu.getMenuTemplate(debugMode);
+    if (menuTemplate.length > 0) {
+      electron.Menu.setApplicationMenu(electron.Menu.buildFromTemplate(menuTemplate));
+    }
+
+    // Register a custom protocol so we can use absolute paths in the web app.
+    // This also acts as a kind of chroot for the web app, so it cannot access
+    // the user's filesystem (important for the DigitalOcean pages we customise).
+    // Hostnames are ignored.
+    electron.protocol.registerFileProtocol(
+        'outline',
+        (request, callback) => {
+          const appPath = new url.URL(request.url).pathname;
+          const filesystemPath = path.join(__dirname, 'server_manager/web_app', appPath);
+          callback(filesystemPath);
+        },
+        (error) => {
+          if (error) {
+            throw new Error('Failed to register outline protocol');
+          }
+        });
+    mainWindow = createMainWindow();
+  });
+
+  ipcMain.on('app-ui-ready', () => {
+    // Check for updates after the UI is loaded; otherwise the UI may miss the
+    //'update-downloaded' event.
+    if (!debugMode) {
+      autoUpdater.checkForUpdates();
+    }
+  });
+  const UPDATE_DOWNLOADED_EVENT = 'update-downloaded';
+  autoUpdater.on(UPDATE_DOWNLOADED_EVENT, (ev, info) => {
+    if (!!mainWindow) {
+      mainWindow.webContents.send(UPDATE_DOWNLOADED_EVENT);
+    }
+  });
+
+  // Set of fingerprints.  All values are true.
```

Comment:
The `All values are true` comment is vestigial (from when this was using an Object, instead of a Set).

Reply:
Removed